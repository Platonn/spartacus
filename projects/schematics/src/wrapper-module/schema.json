{
  "$schema": "http://json-schema.org/schema",
  "$id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "featureModuleName": {
      "type": "string",
      "description": "Name of the feature module for which to generate the wrapper module."
    }
  },
  "required": ["featureModuleName"]
}
