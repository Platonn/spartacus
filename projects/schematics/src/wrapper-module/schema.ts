import { ExecutionOptions } from '@angular-devkit/schematics';

export interface Schema extends Partial<ExecutionOptions> {
  project: string;
  /**
   * The marker module for which
   * to find or generate the wrapper module.
   */
  markerModuleName: string;
  /**
   * Name of the feature module
   * which to add to the wrapper module.
   */
  featureModuleName: string;
}
