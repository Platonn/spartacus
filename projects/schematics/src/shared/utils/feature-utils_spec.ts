import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { Schema as SpartacusOptions } from '../../add-spartacus/schema';
import { UTF_8 } from '../constants';
import { USER_ACCOUNT_SCHEMATICS_CONFIG } from '../lib-configs/user-schematics-config';
import { CLI_USER_ACCOUNT_FEATURE } from '../libs-constants';
import { addFeatures, FeatureConfigurationOverrides } from './feature-utils';
import { LibraryOptions } from './lib-utils';
import { userFeatureModulePath } from './test-utils';

describe('Feature utils', () => {
  const schematicRunner = new SchematicTestRunner(
    'schematics',
    path.join(__dirname, '../../collection.json')
  );

  let appTree: Tree;

  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };

  const spartacusDefaultOptions: SpartacusOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [],
  };

  const BASE_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    lazy: true,
  };

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync(
        'add-spartacus',
        { ...spartacusDefaultOptions, name: 'schematics-test' },
        appTree
      )
      .toPromise();
  });

  describe('addFeatures', () => {
    it('should generate feature modules for the given array of features', async () => {
      appTree = await schematicRunner
        .callRule(
          addFeatures(BASE_OPTIONS, [CLI_USER_ACCOUNT_FEATURE]),
          appTree
        )
        .toPromise();

      expect(
        appTree.read(userFeatureModulePath)?.toString(UTF_8)
      ).toMatchSnapshot();
    });

    it('should override the options', async () => {
      const overrides: Record<string, FeatureConfigurationOverrides> = {
        [CLI_USER_ACCOUNT_FEATURE]: {
          options: {
            ...BASE_OPTIONS,
            lazy: false,
          },
        },
      };

      appTree = await schematicRunner
        .callRule(
          addFeatures(BASE_OPTIONS, [CLI_USER_ACCOUNT_FEATURE], overrides),
          appTree
        )
        .toPromise();

      expect(
        appTree.read(userFeatureModulePath)?.toString(UTF_8)
      ).toMatchSnapshot();
    });

    it('should override the schematics config', async () => {
      const overrides: Record<string, FeatureConfigurationOverrides> = {
        [CLI_USER_ACCOUNT_FEATURE]: {
          schematics: {
            ...USER_ACCOUNT_SCHEMATICS_CONFIG,
            folderName: 'account',
          },
        },
      };

      appTree = await schematicRunner
        .callRule(
          addFeatures(BASE_OPTIONS, [CLI_USER_ACCOUNT_FEATURE], overrides),
          appTree
        )
        .toPromise();

      expect(
        appTree
          .read('src/app/spartacus/features/account/user-feature.module.ts')
          ?.toString(UTF_8)
      ).toMatchSnapshot();
    });
  });
});
