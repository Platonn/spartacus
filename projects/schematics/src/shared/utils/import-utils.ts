import {
  ArrowFunction,
  CallExpression,
  Identifier,
  ImportDeclaration,
  PropertyAccessExpression,
  SourceFile,
  ts as tsMorph,
} from 'ts-morph';
import { CORE_SPARTACUS_SCOPES, SPARTACUS_SCOPE } from '../libs-constants';
import { getSpartacusProviders } from './config-utils';
import { Import } from './new-module-utils';

// TODO:#test
// TODO:#comment
export function isImportedFromSpartacusLibs(
  node: Identifier | string
): boolean {
  return isImportedFrom(node, SPARTACUS_SCOPE);
}

// TODO:#test
// TODO:#comment
export function isImportedFromSpartacusCoreLib(
  node: Identifier | string
): boolean {
  for (const coreScope of CORE_SPARTACUS_SCOPES) {
    if (isImportedFrom(node, coreScope)) {
      return true;
    }
  }

  return false;
}

export function isImportedFrom(
  node: Identifier | string,
  toCheck: string
): boolean {
  let moduleImportPath: string;
  if (typeof node === 'string') {
    moduleImportPath = node;
  } else {
    moduleImportPath = getImportPath(node) ?? '';
  }

  return moduleImportPath.startsWith(toCheck);
}

export function getImportPath(node: Identifier): string | undefined {
  const declaration = getImportDeclaration(node);
  if (declaration) {
    return declaration.getModuleSpecifierValue();
  }

  return undefined;
}

export function getImportDeclaration(
  node: Identifier
): ImportDeclaration | undefined {
  const references = node.findReferencesAsNodes();
  for (const reference of references) {
    const importDeclaration = reference?.getFirstAncestorByKind(
      tsMorph.SyntaxKind.ImportDeclaration
    );
    if (importDeclaration) {
      return importDeclaration;
    }
  }

  return undefined;
}

/**
 * Collects the higher-order arrow functions.
 * E.g. `() => import('@spartacus/cart/base/components/add-to-cart').then((m) => m.AddToCartModule)`,
 * but not the inner one `(m) => m.AddToCartModule`.
 */
export function collectDynamicImports(source: SourceFile): ArrowFunction[] {
  const providers = getSpartacusProviders(source, false);

  let arrowFunctions: ArrowFunction[] = [];
  for (const element of providers) {
    const higherArrowFunctions = element
      .getDescendantsOfKind(tsMorph.SyntaxKind.ArrowFunction)
      .filter((arrowFn) =>
        arrowFn.getParentIfKind(tsMorph.SyntaxKind.PropertyAssignment)
      );
    arrowFunctions = arrowFunctions.concat(higherArrowFunctions);
  }

  return arrowFunctions;
}

/**
 * Returns the import path of the dynamic import (if any).
 * E.g. for the given `() => import('@spartacus/cart/base').then((m) => m.CartBaseModule)` it returns TODO:#schematics - finish
 */
//  TODO:#schematics - test
export function getDynamicImportCallExpression(
  arrowFunction: ArrowFunction
): CallExpression | undefined {
  return arrowFunction
    .getFirstDescendantByKind(tsMorph.SyntaxKind.ImportKeyword)
    ?.getParentIfKind(tsMorph.SyntaxKind.CallExpression);
}

/**
 * Returns the import module of the dynamic import (if any).
 * E.g. for the given `() => import('@spartacus/cart/base').then((m) => m.CartBaseModule)` it returns TODO:#schematics - finish
 */
//  TODO:#schematics - test
export function getDynamicImportPropertyAccess(
  arrowFunction: ArrowFunction
): PropertyAccessExpression | undefined {
  return arrowFunction
    .getFirstDescendantByKind(tsMorph.SyntaxKind.ArrowFunction)
    ?.getFirstDescendantByKind(tsMorph.SyntaxKind.PropertyAccessExpression);
}

//  TODO:#schematics - test
//  TODO:#schematics - comment
export function getDynamicImportImportPath(
  arrowFunction: ArrowFunction
): string | undefined {
  return getDynamicImportCallExpression(arrowFunction)
    ?.getFirstDescendantByKind(tsMorph.SyntaxKind.StringLiteral)
    ?.getLiteralValue();
}

// TODO:#schematics - test
// TODO:#schematics - comment
export function createImports(
  sourceFile: SourceFile,
  imports: Import | Import[]
): ImportDeclaration[] {
  const importDeclarations: ImportDeclaration[] = [];
  ([] as Import[]).concat(imports).forEach((specifiedImport) => {
    const importDeclaration = sourceFile.addImportDeclaration({
      moduleSpecifier: specifiedImport.moduleSpecifier,
      namedImports: specifiedImport.namedImports,
    });
    importDeclarations.push(importDeclaration);
  });

  return importDeclarations;
}

/**
 * Searches through the given module's imports.
 * It first checks the import paths, and if
 * it finds the match, it continues to check if
 * the named imports is contained in the given config.
 */
// TODO:#schematics - test
export function findImport(
  featureModule: SourceFile,
  importPathToFind: string,
  moduleNameToFind: string
): boolean {
  const importDeclarations = featureModule.getImportDeclarations();
  for (const importDeclaration of importDeclarations) {
    const importPath = importDeclaration.getModuleSpecifierValue();
    if (importPathToFind === importPath) {
      const namedImports =
        importDeclaration.getImportClause()?.getNamedImports() ?? [];
      for (const namedImport of namedImports) {
        if (namedImport.getName() === moduleNameToFind) {
          return true;
        }
      }
    }
  }

  return false;
}

// TODO:#schematics - test
// TODO:#schematics - comment
export function isRelative(path: string): boolean {
  return path.startsWith('./') || path.startsWith('../');
}
