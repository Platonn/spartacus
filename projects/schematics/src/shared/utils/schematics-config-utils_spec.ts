/// <reference types="jest" />

import {
  CLI_CHECKOUT_BASE_FEATURE,
  CLI_DIGITAL_PAYMENTS_FEATURE,
} from '../libs-constants';
import { getConfiguredDependencies } from './schematics-config-utils';

describe('schematics-config-util', () => {
  describe('getConfiguredDependencies', () => {
    it('should return the correct feature dependencies for the given feature', () => {
      const result = getConfiguredDependencies(CLI_DIGITAL_PAYMENTS_FEATURE);
      expect(result).toEqual([CLI_CHECKOUT_BASE_FEATURE]);
    });
  });
});
