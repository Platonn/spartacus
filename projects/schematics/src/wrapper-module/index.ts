import {
  chain,
  noop,
  Rule,
  SchematicContext,
  SchematicsException,
  Tree,
} from '@angular-devkit/schematics';
import { ArrowFunction, Node, SourceFile } from 'ts-morph';
import {
  featureFeatureModuleMapping,
  featureSchematicConfigMapping,
  getKeyByMappingValueOrThrow,
} from '../shared/updateable-constants';
import {
  findDynamicImport,
  getModuleConfig,
  isFeatureModule,
  isWrapperModule,
} from '../shared/utils/feature-utils';
import {
  getDynamicImportCallExpression,
  getDynamicImportPropertyAccess,
} from '../shared/utils/import-utils';
import {
  createSpartacusFeatureFolderPath,
  createSpartacusWrapperModuleFileName,
  FeatureConfig,
} from '../shared/utils/lib-utils';
import {
  addModuleImport,
  ensureModuleExists,
  getModulePropertyInitializer,
} from '../shared/utils/new-module-utils';
import { createProgram, saveAndFormat } from '../shared/utils/program';
import { getProjectTsConfigPaths } from '../shared/utils/project-tsconfig-paths';
import { Schema as SpartacusWrapperOptions } from './schema';

// TODO:#schematics - update tsmorph to v14

function createWrapperModule(
  options: SpartacusWrapperOptions,
  featureConfig: FeatureConfig
): Rule {
  return (tree: Tree, _context: SchematicContext) => {
    const basePath = process.cwd();
    const { buildPaths } = getProjectTsConfigPaths(tree, options.project);

    const path = createSpartacusFeatureFolderPath(featureConfig.folderName);
    const name = createSpartacusWrapperModuleFileName(featureConfig.moduleName);

    const rules: Rule[] = [];
    for (const tsconfigPath of buildPaths) {
      const { appSourceFiles } = createProgram(tree, basePath, tsconfigPath);

      for (const sourceFile of appSourceFiles) {
        if (
          isWrapperModule(sourceFile, options.featureModuleName, featureConfig)
        ) {
          // nothing to do if the wrapper module already exists
          return noop();
        }

        if (
          !isFeatureModule(sourceFile, options.featureModuleName, featureConfig)
        ) {
          continue;
        }
        rules.push(
          ensureModuleExists({
            path,
            name,
            project: options.project,
            /**
             * Only temporarily import the wrapper module to the feature module.
             * Will be removed once we start updating the wrapper module.
             *
             * This is a workaround for a weird behavior of the ts-morph library,
             * which does not "see" the newly created module if it is not
             * referenced anywhere.
             */
            module: sourceFile.getBaseNameWithoutExtension(),
          })
        );
        break;
      }
    }

    return chain(rules);
  };
}

function updateWrapperModule(
  options: SpartacusWrapperOptions,
  featureConfig: FeatureConfig
): Rule {
  return (tree: Tree, _context: SchematicContext) => {
    const basePath = process.cwd();
    const { buildPaths } = getProjectTsConfigPaths(tree, options.project);

    const featureModuleConfig = getModuleConfig(
      options.featureModuleName,
      featureConfig
    );
    if (!featureModuleConfig) {
      // TODO:#schematics - do what?
      return noop();
    }

    const path = createSpartacusFeatureFolderPath(featureConfig.folderName);
    const name = createSpartacusWrapperModuleFileName(featureConfig.moduleName);

    const rules: Rule[] = [];
    for (const tsconfigPath of buildPaths) {
      const { appSourceFiles } = createProgram(tree, basePath, tsconfigPath);

      for (const wrapperModule of appSourceFiles) {
        if (!wrapperModule.getFilePath().includes(`${path}/${name}`)) {
          continue;
        }

        // TODO:#schematics - order?
        addModuleImport(wrapperModule, {
          import: {
            moduleSpecifier: featureModuleConfig.importPath,
            namedImports: [featureModuleConfig.name],
          },
          content: featureModuleConfig.name,
        });
        saveAndFormat(wrapperModule);
      }
    }

    return chain(rules);
  };
}

function updateFeatureModule(
  options: SpartacusWrapperOptions,
  featureConfig: FeatureConfig
): Rule {
  return (tree: Tree, _context: SchematicContext) => {
    const basePath = process.cwd();
    const { buildPaths } = getProjectTsConfigPaths(tree, options.project);

    const featureModuleConfig = getModuleConfig(
      options.featureModuleName,
      featureConfig
    );
    if (!featureModuleConfig) {
      // TODO:#schematics - how to handle it?
      return noop();
    }

    const path = createSpartacusFeatureFolderPath(featureConfig.folderName);
    const name = createSpartacusWrapperModuleFileName(featureConfig.moduleName);

    const rules: Rule[] = [];
    for (const tsconfigPath of buildPaths) {
      const { appSourceFiles } = createProgram(tree, basePath, tsconfigPath);

      for (const wrapperModule of appSourceFiles) {
        if (!wrapperModule.getFilePath().includes(`${path}/${name}`)) {
          continue;
        }

        const featureModule = findFeatureModule(wrapperModule);
        if (!featureModule) {
          // TODO:#schematics - how to handle it?
          continue;
        }

        const dynamicImport = findDynamicImport(
          featureModule,
          featureModuleConfig.importPath,
          featureModuleConfig.name
        );
        if (!dynamicImport) {
          // TODO:#schematics - how to handle it?
          continue;
        }

        const wrapperModuleClassName =
          wrapperModule.getClasses()[0].getName() ?? '';
        updateDynamicImportPath(
          dynamicImport,
          featureModule.getRelativePathAsModuleSpecifierTo(
            wrapperModule.getFilePath()
          )
        );
        updateDynamicImportModuleName(dynamicImport, wrapperModuleClassName);

        // remove the dummy import
        const ngImports = getModulePropertyInitializer(
          featureModule,
          'imports',
          false
        );
        if (!ngImports) {
          continue;
        }
        for (const element of ngImports.getElements()) {
          if (element.getText() === wrapperModuleClassName) {
            ngImports.removeElement(element);
            break;
          }
        }

        saveAndFormat(featureModule);
      }
    }

    return chain(rules);
  };
}

function findFeatureModule(wrapperModule: SourceFile): SourceFile | undefined {
  const referenceSymbols = wrapperModule.getClasses()[0].findReferences();
  for (const referenceSymbol of referenceSymbols) {
    for (const reference of referenceSymbol.getReferences()) {
      const node = reference.getNode();
      const parent = node.getParentOrThrow();
      if (!Node.isArrayLiteralExpression(parent)) {
        continue;
      }

      return parent.getSourceFile();
    }
  }

  return undefined;
}

function updateDynamicImportPath(
  dynamicImport: ArrowFunction,
  path: string
): void {
  getDynamicImportCallExpression(dynamicImport)
    ?.removeArgument(0)
    ?.insertArgument(0, `'${path}'`);
}

function updateDynamicImportModuleName(
  dynamicImport: ArrowFunction,
  wrapperModuleName: string
): void {
  getDynamicImportPropertyAccess(dynamicImport)?.replaceWithText(
    `m.${wrapperModuleName}`
  );
}

/**
 * TODO:#schematics - test
 * 1. multiple dynamic imports in one file
 * 2. mixed dynamic imports to local and node_modules files
 * 3. Run e.g DP and expect all previous extensions to be added to the wrapper module
 */

/**
 * Generates wrapper modules for the given
 * Spartacus feature module.
 */
export function generateWrapperModule(options: SpartacusWrapperOptions): Rule {
  return (_tree: Tree, _context: SchematicContext): Rule => {
    const feature = getKeyByMappingValueOrThrow(
      featureFeatureModuleMapping,
      options.featureModuleName
    );
    const featureConfig = featureSchematicConfigMapping.get(feature);
    if (!featureConfig) {
      throw new SchematicsException(
        `Config not found for the given feature '${feature}'`
      );
    }

    return chain([
      createWrapperModule(options, featureConfig),
      updateWrapperModule(options, featureConfig),
      updateFeatureModule(options, featureConfig),
    ]);
  };
}

// TODO:#schematics - algo
// pull the schematics config associated with the given feature module name
// iterate through the collected feature modules
// check for imports of the given feature module
// if found, make sure it's actually imported in the ngmodule's imports[]
// if not found, fall back to checking the dynamic imports import in providers array
// if the dynamic import is found, check the import path
// ---
// if found in the dynamic import --> tear it down and create a wrapper module
// if found in the ngmodule's imports[] --> we need to check if it's not already a wrapper module.
// we can do that by checking for the Module's references, and check if there's a reference to the module using a dynamic import.
// if yes, we found a wrapper module, and we can just import the given module name to it
// and sort its imports[].
// if there isn't a dynamic import reference, what are we working with? An eagerly loaded feature?
