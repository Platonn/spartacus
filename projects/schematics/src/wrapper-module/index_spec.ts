import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { Schema as SpartacusOptions } from '../add-spartacus/schema';
import { CART_BASE_MODULE } from '../shared/lib-configs/cart-schematics-config';
import { CHECKOUT_BASE_MODULE } from '../shared/lib-configs/checkout-schematics-config';
import { CLI_CHECKOUT_BASE_FEATURE } from '../shared/libs-constants';
import { LibraryOptions } from '../shared/utils/lib-utils';
import { createProgram } from '../shared/utils/program';
import { getProjectTsConfigPaths } from '../shared/utils/project-tsconfig-paths';
import {
  cartBaseFeatureModulePath,
  checkoutFeatureModulePath,
  checkoutWrapperModulePath,
} from '../shared/utils/test-utils';
import { Schema as SpartacusWrapperOptions } from '../wrapper-module/schema';

const collectionPath = path.join(__dirname, '../collection.json');

describe('Spartacus Wrapper Module Schematics: ng g @spartacus/schematics:wrapper-module', () => {
  const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

  let appTree: Tree;
  let buildPath: string;

  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };

  const defaultOptions: SpartacusOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [CLI_CHECKOUT_BASE_FEATURE],
  };

  const BASE_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    lazy: true,
  };

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync(
        'ng-add',
        { ...defaultOptions, name: 'schematics-test' },
        appTree
      )
      .toPromise();

    buildPath = getProjectTsConfigPaths(appTree, BASE_OPTIONS.project)
      .buildPaths[0];
  });

  describe('Checkout - One dynamic import in the file', () => {
    it('should generate appropriate feature module', async () => {
      const checkoutOptions: SpartacusWrapperOptions = {
        project: 'schematics-test',
        featureModuleName: CHECKOUT_BASE_MODULE,
      };
      appTree = await schematicRunner
        .runSchematicAsync('wrapper-module', checkoutOptions, appTree)
        .toPromise();

      const { program } = createProgram(appTree, appTree.root.path, buildPath);

      const checkoutFeaturesModule = program.getSourceFileOrThrow(
        checkoutFeatureModulePath
      );
      const checkoutWrapperModule = program.getSourceFileOrThrow(
        checkoutWrapperModulePath
      );

      expect(checkoutFeaturesModule.print()).toMatchSnapshot();
      expect(checkoutWrapperModule.print()).toMatchSnapshot();
    });
  });

  describe('Cart - multiple dynamic imports in the file', () => {
    it('should generate appropriate feature module', async () => {
      const checkoutOptions: SpartacusWrapperOptions = {
        project: 'schematics-test',
        featureModuleName: CART_BASE_MODULE,
      };
      appTree = await schematicRunner
        .runSchematicAsync('wrapper-module', checkoutOptions, appTree)
        .toPromise();

      const { program } = createProgram(appTree, appTree.root.path, buildPath);

      const cartFeaturesModule = program.getSourceFileOrThrow(
        cartBaseFeatureModulePath
      );
      const cartWrapperModule = program.getSourceFileOrThrow(
        'src/app/spartacus/features/cart/cart-base-wrapper.module.ts'
      );

      expect(cartFeaturesModule.print()).toMatchSnapshot();
      expect(cartWrapperModule.print()).toMatchSnapshot();
    });
  });

  describe('double execution', () => {
    it('should not change anything', async () => {
      const checkoutOptions: SpartacusWrapperOptions = {
        project: 'schematics-test',
        featureModuleName: 'CheckoutModule',
      };
      appTree = await schematicRunner
        .runSchematicAsync('wrapper-module', checkoutOptions, appTree)
        .toPromise();
      appTree = await schematicRunner
        .runSchematicAsync('wrapper-module', checkoutOptions, appTree)
        .toPromise();

      const { program } = createProgram(appTree, appTree.root.path, buildPath);

      const checkoutFeaturesModule = program.getSourceFileOrThrow(
        checkoutFeatureModulePath
      );
      const checkoutWrapperModule = program.getSourceFileOrThrow(
        checkoutWrapperModulePath
      );

      expect(checkoutFeaturesModule.print()).toMatchSnapshot();
      expect(checkoutWrapperModule.print()).toMatchSnapshot();
    });
  });
});
