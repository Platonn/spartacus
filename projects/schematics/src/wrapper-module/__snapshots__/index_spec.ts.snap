// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spartacus Wrapper Module Schematics: ng g @spartacus/schematics:wrapper-module Cart - multiple dynamic imports in the file should generate appropriate feature module 1`] = `
"import { NgModule } from '@angular/core';
import { cartBaseTranslationChunksConfig, cartBaseTranslations } from \\"@spartacus/cart/base/assets\\";
import { ADD_TO_CART_FEATURE, CartBaseRootModule, CART_BASE_FEATURE, MINI_CART_FEATURE } from \\"@spartacus/cart/base/root\\";
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
@NgModule({
    declarations: [],
    imports: [
        CartBaseRootModule
    ],
    providers: [provideConfig(<CmsConfig>{
            featureModules: {
                [CART_BASE_FEATURE]: {
                    module: () => import('./cart-base-wrapper.module').then((m) => m.CartBaseWrapperModule),
                }, [MINI_CART_FEATURE]: {
                    module: () => import('@spartacus/cart/base/components/mini-cart').then((m) => m.MiniCartModule),
                }, [ADD_TO_CART_FEATURE]: {
                    module: () => import('@spartacus/cart/base/components/add-to-cart').then((m) => m.AddToCartModule),
                },
            }
        }),
        provideConfig(<I18nConfig>{
            i18n: {
                resources: cartBaseTranslations,
                chunks: cartBaseTranslationChunksConfig,
            },
        })
    ]
})
export class CartBaseFeatureModule {
}
"
`;

exports[`Spartacus Wrapper Module Schematics: ng g @spartacus/schematics:wrapper-module Cart - multiple dynamic imports in the file should generate appropriate feature module 2`] = `
"import { NgModule } from '@angular/core';
import { CartBaseModule } from \\"@spartacus/cart/base\\";
@NgModule({
    declarations: [],
    imports: [
        CartBaseModule
    ]
})
export class CartBaseWrapperModule {
}
"
`;

exports[`Spartacus Wrapper Module Schematics: ng g @spartacus/schematics:wrapper-module Checkout - One dynamic import in the file should generate appropriate feature module 1`] = `
"import { NgModule } from '@angular/core';
import { checkoutTranslationChunksConfig, checkoutTranslations } from \\"@spartacus/checkout/base/assets\\";
import { CheckoutRootModule, CHECKOUT_FEATURE } from \\"@spartacus/checkout/base/root\\";
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
@NgModule({
    declarations: [],
    imports: [
        CheckoutRootModule
    ],
    providers: [provideConfig(<CmsConfig>{
            featureModules: {
                [CHECKOUT_FEATURE]: {
                    module: () => import('./checkout-wrapper.module').then((m) => m.CheckoutWrapperModule),
                },
            }
        }),
        provideConfig(<I18nConfig>{
            i18n: {
                resources: checkoutTranslations,
                chunks: checkoutTranslationChunksConfig,
            },
        })
    ]
})
export class CheckoutFeatureModule {
}
"
`;

exports[`Spartacus Wrapper Module Schematics: ng g @spartacus/schematics:wrapper-module Checkout - One dynamic import in the file should generate appropriate feature module 2`] = `
"import { NgModule } from '@angular/core';
import { CheckoutModule } from \\"@spartacus/checkout/base\\";
@NgModule({
    declarations: [],
    imports: [
        CheckoutModule
    ]
})
export class CheckoutWrapperModule {
}
"
`;

exports[`Spartacus Wrapper Module Schematics: ng g @spartacus/schematics:wrapper-module double execution should not change anything 1`] = `
"import { NgModule } from '@angular/core';
import { checkoutTranslationChunksConfig, checkoutTranslations } from \\"@spartacus/checkout/base/assets\\";
import { CheckoutRootModule, CHECKOUT_FEATURE } from \\"@spartacus/checkout/base/root\\";
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
@NgModule({
    declarations: [],
    imports: [
        CheckoutRootModule
    ],
    providers: [provideConfig(<CmsConfig>{
            featureModules: {
                [CHECKOUT_FEATURE]: {
                    module: () => import('./checkout-wrapper.module').then((m) => m.CheckoutWrapperModule),
                },
            }
        }),
        provideConfig(<I18nConfig>{
            i18n: {
                resources: checkoutTranslations,
                chunks: checkoutTranslationChunksConfig,
            },
        })
    ]
})
export class CheckoutFeatureModule {
}
"
`;

exports[`Spartacus Wrapper Module Schematics: ng g @spartacus/schematics:wrapper-module double execution should not change anything 2`] = `
"import { NgModule } from '@angular/core';
import { CheckoutModule } from \\"@spartacus/checkout/base\\";
@NgModule({
    declarations: [],
    imports: [
        CheckoutModule
    ]
})
export class CheckoutWrapperModule {
}
"
`;
